chainerx_cc/chainerx/routines/manipulation.cc

```diff
@@ -763,4 +764,46 @@ Array VStack(const std::vector<Array>& arrays) {
     return Concatenate(reshaped_arrays, 0);
 }
 
+Array Moveaxis(const Array& a, const OptionalAxes& src, const OptionalAxes& dst) {
+    Axes real_src;
+    Axes real_dst;
+    if (src.has_value()) {
+        real_src = internal::GetNormalizedAxes(*src, a.ndim());
+    }
+    if (dst.has_value()) {
+        real_dst = internal::GetNormalizedAxes(*dst, a.ndim());
+    }
+
+    if (real_src.size() != real_dst.size()) {
+        throw DimensionError{"Invalid Source or Destination Axes"};
+    }
+
+    if (real_src.size() == 0) {
+        return a;
+    }
+
+    Axes order;
+    std::set<int8_t> src_set(real_src.begin(), real_src.end());
+    for (int8_t i = 0; i < a.ndim(); ++i) {
+        if (src_set.find(i) == src_set.end()) {
+            order.emplace_back(i);
+        }
+    }
+
+    std::vector<std::pair<int8_t, int8_t>> dst_src_pairs;
+    for (int8_t i = 0; i < real_dst.ndim(); i++) {
+        dst_src_pairs.emplace_back(std::make_pair(real_dst[i], real_src[i]));
+    }
+
+    std::sort(dst_src_pairs.begin(), dst_src_pairs.end(), [](std::pair<int8_t, int8_t> a, std::pair<int8_t, int8_t> b) {
+        return a.first < b.first;
+    });
+
+    for (uint8_t i = 0; i < dst_src_pairs.size(); i++) {
+        order.insert(order.begin() + dst_src_pairs[i].first, dst_src_pairs[i].second);
+    }
```

Comment:
How about implementing as follows for the performance and the readability?
```cpp
Axes order(a.ndim()), src_axes(a.ndim()), dst_axes(a.ndim());
std::iota(src_axes.begin(), src_axes.end(), 0);
std::iota(dst_axes.begin(), dst_axes.end(), 0);

for (uint8_t i = 0; i < src.ndim(); ++i) {
    order[dst[i]] = src[i];
    src_axes[src[i]] = -1;
    dst_axes[dst[i]] = -1;
}

std::remove(src_axes.begin(), src_axes.end(), -1);
std::remove(dst_axes.begin(), dst_axes.end(), -1);
CHAINERX_ASSERT(src_axes.size() == a.ndim() - src.ndim());
CHAINERX_ASSERT(dst_axes.size() == a.ndim() - src.ndim());

for (uint8_t i = 0; a.ndim() - src.ndim(); ++i) {
    order[dst_axes[i]] = src_axes[i];
}
```

Reply:
Realllly nice. Thank You. Have done this with slight changes as needed. Please review.

Reply:
`source_axes.erase` and `destination_axes.erase` seem needed only for `CHAINERX_ASSERT`. I want to avoid these `erase` because `CHAINERX_ASSERT` is operated only in debug mode.
```cpp
std::remove(source_axes.begin(), source_axes.end(), -1);
std::remove(destination_axes.begin(), destination_axes.end(), -1);
CHAINERX_ASSERT(static_cast<int8_t>(std::find(source_axes.begin(), source_axes.end(), -1) - source_axes.begin()) == a.ndim() - source.ndim());
CHAINERX_ASSERT(static_cast<int8_t>(std::find(destination_axes.begin(), destination_axes.end(), -1) - destination_axes.begin()) == a.ndim() - destination.ndim());
```

Reply:
We wont need `std::find` as `std::remove` will return the `iterator` with valid position. 
```CPP
Axes::iterator src_iterator = std::remove(source_axes.begin(), source_axes.end(), -1);
Axes::iterator dst_iterator = std::remove(destination_axes.begin(), destination_axes.end(), -1);
CHAINERX_ASSERT(static_cast<int8_t>(src_iterator - source_axes.begin()) == a.ndim() - source.ndim());
CHAINERX_ASSERT(static_cast<int8_t>(dst_iterator - destination_axes.begin()) == a.ndim() - destination.ndim());
```

Also we will need to update the `for` loop below this to consume those iterators.

Reply:
In release mode, `CHAINERX_ASSERT` is not operated, so we don't have to (and don't want to) shrink these axes.

Reply:
Sorry, I got it wrong.